### ISSUE TITLE
SQLAlchemy QueuePool limit TimeoutError

### ISSUE THREAD

**Issue by jreinhar3dt** [Flask-Middleware/flask-security#859]
This is my first project with this library and I am facing issues that the database reaches a timeout (note that I've set the pool size to 1 for debugging, but also face this problem with the default values)
```
sqlalchemy.exc.TimeoutError: QueuePool limit of size 1 overflow 0 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/20/3o7r)
```

It seems that within flask-security the connection is not closed/the transaction committed during the login, since the error can be seen after calling the `login` endpoint repeatedly. I don't know if this is a misconfiguration or how to inject a fix that does that.

I followed the example https://flask-security-too.readthedocs.io/en/stable/quickstart.html#basic-sqlalchemy-application-with-session and can replicate this behavior with that example. I use `Flask==2.2.5`, `flask-security-too==5.1.2` and `sqlalchemy==2.0.21`, I've also tried to use some other versions but to no avail.

**Comment by jreinhar3dt** [comment]
It seems that using `sqlalchemy==1.4.49` helps with this issue, however I am not sure if that version uses a QueuePool as it throws an error stating that setting the pool size is not a valid argument.

This version then shows errors about SQLite objects access from different threads, but that can be circumvented with `, connect_args={'check_same_thread': False}`.

However I wonder what steps would be necessary to use the newer version of SQLAlchemy.

**Comment by jwag956** [comment]
Interesting - there is a lot of info about pools at the sqlalchemy site:
this section on fork: https://docs.sqlalchemy.org/en/20/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork
might be interesting as well as this on logging:
https://docs.sqlalchemy.org/en/20/core/pooling.html#logging-reset-on-return-events which you might be able to use to track connections and see if you can get any idea who isn't returning them!

It would be  useful to see how you are initializing your DB, and what kind of test setup (what front end, what options to wsgi etc.)

**Comment by jreinhar3dt** [comment]
Thanks for the information, I've included the pool logging and the output for the `login` call looks like
```
023-10-18 10:49:33,642 DEBUG sqlalchemy.pool.impl.QueuePool Created new connection <sqlite3.Connection object at 0x00000186498274E0>
2023-10-18 10:49:33,642 DEBUG sqlalchemy.pool.impl.QueuePool Connection <sqlite3.Connection object at 0x00000186498274E0> checked out from pool
127.0.0.1 - - [18/Oct/2023 10:49:33] "POST //login HTTP/1.1" 400 -
```
indicating that the connection is checked out but not returned while handling the request. I did found out that they are returned after some time
```
2023-10-18 10:50:53,975 DEBUG sqlalchemy.pool.impl.QueuePool Connection <sqlite3.Connection object at 0x00000186498274E0> being returned to pool
2023-10-18 10:50:53,976 DEBUG sqlalchemy.pool.impl.QueuePool Connection <sqlite3.Connection object at 0x00000186498274E0> rollback-on-return
```
but it seems that this does not work reliable, since after some calls the endpoint always returns `500` because of the `sqlalchemy.exc.TimeoutError` caused by the pool limit.

What I would have expected is that after each login attempt the, connection would be returned immediately immediately, like it happens in the setup of the database
```
2023-10-18 10:48:21,792 DEBUG sqlalchemy.pool.impl.QueuePool Created new connection <sqlite3.Connection object at 0x000001EE2574AB70>
2023-10-18 10:48:21,792 DEBUG sqlalchemy.pool.impl.QueuePool Connection <sqlite3.Connection object at 0x000001EE2574AB70> checked out from pool 
2023-10-18 10:48:21,794 DEBUG sqlalchemy.pool.impl.QueuePool Connection <sqlite3.Connection object at 0x000001EE2574AB70> being returned to pool
2023-10-18 10:48:21,794 DEBUG sqlalchemy.pool.impl.QueuePool Connection <sqlite3.Connection object at 0x000001EE2574AB70> rollback-on-return    
2023-10-18 10:48:21,798 DEBUG sqlalchemy.pool.impl.QueuePool Connection <sqlite3.Connection object at 0x000001EE2574AB70> checked out from pool 
2023-10-18 10:48:21,800 DEBUG sqlalchemy.pool.impl.QueuePool Connection <sqlite3.Connection object at 0x000001EE2574AB70> being returned to pool
2023-10-18 10:48:21,800 DEBUG sqlalchemy.pool.impl.QueuePool Connection <sqlite3.Connection object at 0x000001EE2574AB70> rollback-on-return 
```

Regarding the setup, I am using this example https://flask-security-too.readthedocs.io/en/stable/quickstart.html#basic-sqlalchemy-application-with-session basically unmodified except for `engine = create_engine('sqlite:////tmp/test.db', pool_size=2, max_overflow=1, echo_pool="debug")`, use just `py app.py` to run the app and do not use a front-end but call the `login` endpoint with Postman. Not sure if this is relevant but I am using Windows 10.

**Comment by jwag956** [comment]
Thanks for the detailed info - I'll try to reproduce and see what's going on...

**Comment by jreinhar3dt** [comment]
Thanks a lot, let me know if you need additional information. I've noticed that the connections are returned during the shutdown of the application, but that is probably not helpful information.

**Comment by jwag956** [comment]
Ok - this appears to be a very longstanding oversite in the quickstart. It mostly works since IF the request is a success, then a COMMIT is done and the connection appears to be returned - in your case - on an error (you are getting back a 400) the query is done and a connection established - but never closed.

Please try the following - in your code:
`engine = create_engine(app.config["SQLALCHEMY_DATABASE_URI"], **engine_kwargs)
   db_session = scoped_session( sessionmaker(autocommit=False, autoflush=False, bind=engine))
   app.teardown_appcontext(lambda exc: db_session.close())`

Add the last line (app.teardown_appcontext - and see if that  gets things working.

**Comment by jreinhar3dt** [comment]
Thank you! Adding `app.teardown_appcontext(lambda exc: db_session.close())` helped, the connections for failed logins are now returned. Is this a workaround or the intended implementation?

**Comment by jwag956** [comment]
That solution is the required one - it is almost identical to how Flask-SQLAlchemy handles the same issue. When using scoped_sessions directly, the application is responsible for engine creation, and connection management. (which is basically what Flask-SQLALchemy is all about). I am updating the Quickstart guide.

**Comment by jwag956** [comment]
Fixed by pr #863
